codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```diff
@@ -0,0 +1,835 @@
+
+            w.writeDocs("update the remaining time");
+            w.write("remainingTime = remainingTime - delay");
+            w.write("");
+
+            w.writeDocs("rewind transport stream");
+            w.write("if rewindable, ok := in.Request.(interface{ RewindStream() error }); ok {");
+            w.write("if err := rewindable.RewindStream(); err != nil {");
+            w.addUseImports(SmithyGoDependency.FMT);
+            w.write("return out, metadata, fmt.Errorf(\"failed to rewind transport stream for retry, %w\", err)\n}");
+            w.write("}");
+            w.write("");
+
+            Symbol sleepWithContextSymbol = SymbolUtils.createValueSymbolBuilder(
+                    "SleepWithContext", SmithyGoDependency.SMITHY_WAITERS
+            ).build();
+            w.writeDocs("sleep for the delay amount before invoking a request");
+            w.openBlock("if err = $T(ctx, delay); err != nil {", "}",
+                    sleepWithContextSymbol, () -> {
+                        w.write("return out, metadata, fmt.Errorf(\"request cancelled while waiting, %w\", err)");
+                    });
+
+            // enable logger
+            w.writeDocs("log attempt");
+            w.openBlock("if m.enableLogger {", "}", () -> {
+                w.write("logger.Logf(logging.Debug, "
+                        + "fmt.Sprintf(\"retrying waiter request, attempt count: %d\", attempt))\n");
+            });
+            w.write("");
+            // end retry only behavior
+            w.write("}");
+
+            w.write("");
+            w.writeDocs("attempt request");
+            w.write("out, metadata, err = next.HandleFinalize(ctx, attemptInput)");
+            w.write("");
+
+            w.writeDocs("check for state mutation");
+            w.write("output := out.Result.($P)", outputSymbol);
+            w.write("retryable, err := m.waiterStateMutator(ctx, m.params, output, err)");
+            w.write("");
+
+            w.openBlock("if !retryable || err != nil {", "}", () -> {
+                w.write("if m.enableLogger {");
+                w.openBlock("if err != nil {", "} else {", () -> {
+                    w.write("logger.Logf(logging.Debug, "
+                            + "\"waiter transitioned to a failed state with unretryable error %v\", err)");
+                });
+                w.write("logger.Logf(logging.Debug, \"waiter transitioned to a success state\")}");
+                w.write("}");
+                w.write("return out, metadata, err");
+            });
+            // end loop
+            w.write("}");
+
+            w.openBlock("if m.enableLogger {", "}", () -> {
+                w.write("logger.Logf(logging.Debug, \"max retry attempts exhausted, max %d\", attempt)");
+            });
+
+            w.addUseImports(SmithyGoDependency.FMT);
+            w.write("return out, metadata, "
+                    + "fmt.Errorf(\"exhausted all retry attempts while waiting for the resource state\")");
+
```

Comment:
is a log message really needed if an error will will duplicate it?

Reply:
Also this section seems to be also returning on the first attempt regardless if retryable or not.

i.e i'd expect a `continue` or similar if `retryable` is true.